src/parsing.jl

```diff
@@ -404,19 +433,74 @@ function pqparse(::Type{Dates.CompoundPeriod}, str::AbstractString)
     return Dates.CompoundPeriod(periods)
 end
 
+
 ## ranges
-_DEFAULT_TYPE_MAP[:int4range] = Interval{Int32}
-_DEFAULT_TYPE_MAP[:int8range] = Interval{Int64}
-_DEFAULT_TYPE_MAP[:numrange] = Interval{Decimal}
-_DEFAULT_TYPE_MAP[:tsrange] = Interval{DateTime}
-_DEFAULT_TYPE_MAP[:tstzrange] = Interval{ZonedDateTime}
-_DEFAULT_TYPE_MAP[:daterange] = Interval{Date}
+_RANGE_TYPE_MAP = Dict{Symbol, Type}()
+_RANGE_TYPE_MAP[:int4range] = Int32
+_RANGE_TYPE_MAP[:int8range] = Int64
+_RANGE_TYPE_MAP[:numrange] = Decimal
+_RANGE_TYPE_MAP[:tsrange] = DateTime
+_RANGE_TYPE_MAP[:tstzrange] = ZonedDateTime
+_RANGE_TYPE_MAP[:daterange] = Date
+
+foreach(x -> _DEFAULT_TYPE_MAP[x] = Interval{_RANGE_TYPE_MAP[x]}, keys(_RANGE_TYPE_MAP))
 
 function pqparse(::Type{Interval{T}}, str::AbstractString) where T
     str == "empty" && return Interval{T}()
     return parse(Interval{T}, str; element_parser=pqparse)
 end
 
+# How to parse range binary fetch is shown here
+# https://github.com/postgres/postgres/blob/31079a4a8e66e56e48bad94d380fa6224e9ffa0d/src/backend/utils/adt/rangetypes.c#L162
+const RANGE_EMPTY =                 0b00000001
+const RANGE_LOWER_BOUND_INCLUSIVE = 0b00000010
+const RANGE_UPPER_BOUND_INCLUSIVE = 0b00000100
+const RANGE_LOWER_BOUND_INFINITIY = 0b00001000
+const RANGE_UPPER_BOUND_INFINITIY = 0b00010000
+const RANGE_LOWER_BOUND_NULL =      0b00100000
+const RANGE_UPPER_BOUND_NULL =      0b01000000
+
+function generate_range_binary_parser(symbol)
+    @eval function Base.parse(::Type{Interval{T}}, pqv::PQBinaryValue{$(oid(symbol))}) where T
+
+        current_pointer = data_pointer(pqv)
+        flags = ntoh(unsafe_load(Ptr{UInt8}(current_pointer)))
+        current_pointer += sizeof(UInt8)
+
+        Bool(flags & RANGE_EMPTY) && return Interval{T}()
+
+        lower_value = nothing
+        lower_bound = Unbounded
+        # if there is a lower bound
+        if iszero(flags & (RANGE_LOWER_BOUND_INFINITIY | RANGE_LOWER_BOUND_NULL))
```

Comment:
This may be faster, as the compiler is sometimes better at reasoning about `===` since it can't be overloaded
```suggestion
        if 0x00 === (flags & (RANGE_LOWER_BOUND_INFINITIY | RANGE_LOWER_BOUND_NULL))
```

Reply:
I don't understand how that could ever have any runtime impacts in this case, where the types are all identical and known.

Reply:
Seems like this case will happily inline to do that anyway,
so this will not help
```
julia> foo(a,b,c) = iszero(a & (b | c))
foo (generic function with 1 method)

julia> @code_typed foo(0x1, 0x2, 0x3)
CodeInfo(
1 ─ %1 = Base.or_int(b, c)::UInt8
│   %2 = Base.and_int(a, %1)::UInt8
│   %3 = (%2 === 0x00)::Bool
└──      return %3
) => Bool
```